category_theory/limits/cones.lean

```diff
@@ -0,0 +1,245 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Stephen Morgan, Scott Morrison
+
+import category_theory.natural_isomorphism
+import category_theory.whiskering
+import category_theory.const
+import category_theory.opposites
+import category_theory.yoneda
+
+universes u u' v
+
+open category_theory
+
+variables {J : Type v} [small_category J]
+variables {C : Type u} [𝒞 : category.{u v} C]
+include 𝒞
+
+open category_theory
+open category_theory.category
+open category_theory.functor
+
+namespace category_theory
+
+namespace functor
+variables {J C} (F : J ⥤ C)
+
+/--
+`F.cones` is the functor assigning to an object `X` the type of
+natural transformations from the constant functor with value `X` to `F`.
+An object representing this functor is a limit of `F`.
+-/
+def cones : Cᵒᵖ ⥤ Type _ := (const (Jᵒᵖ)) ⋙ (op_inv J C) ⋙ (yoneda.obj F)
+
+lemma cones_obj (X : C) : F.cones.obj X = ((const J).obj X ⟹ F) := rfl
+
+/--
+`F.cocones` is the functor assigning to an object `X` the type of
+natural transformations from `F` to the constant functor with value `X`.
+An object corepresenting this functor is a colimit of `F`.
+-/
+def cocones : C ⥤ Type _ := (const J) ⋙ (coyoneda.obj F)
+
+lemma cocones_obj (X : C) : F.cocones.obj X = (F ⟹ (const J).obj X) := rfl
+
+end functor
+
+
+namespace limits
+
+/--
+A `c : cone F` is:
+* an object `c.X` and
+* a natural transformation `c.π : c.X ⟹ F` from the constant `c.X` functor to `F`.
+
+`cone F` is equivalent, in the obvious way, to `Σ X, F.cones.obj X`.
+-/
+structure cone (F : J ⥤ C) :=
+(X : C)
+(π : (const J).obj X ⟹ F)
+
+@[simp] lemma cone.w {F : J ⥤ C} (c : cone F) {j j' : J} (f : j ⟶ j') :
+  c.π.app j ≫ F.map f = c.π.app j' :=
+by convert ←(c.π.naturality f).symm; apply id_comp
+
+/--
+A `c : cocone F` is
+* an object `c.X` and
+* a natural transformation `c.ι : F ⟹ c.X` from `F` to the constant `c.X` functor.
+
+`cocone F` is equivalent, in the obvious way, to `Σ X, F.cocones.obj X`.
+-/
+structure cocone (F : J ⥤ C) :=
+(X : C)
+(ι : F ⟹ (const J).obj X)
+
+@[simp] lemma cocone.w {F : J ⥤ C} (c : cocone F) {j j' : J} (f : j ⟶ j') :
+  F.map f ≫ c.ι.app j' = c.ι.app j :=
+by convert ←(c.ι.naturality f); apply comp_id
+
+
+variables {F : J ⥤ C}
+
+namespace cone
+@[simp] def extensions (c : cone F) : yoneda.obj c.X ⟶ F.cones :=
+{ app := λ X f, ((const J).map f) ≫ c.π }
+
+/-- A map to the vertex of a cone induces a cone by composition. -/
+@[simp] def extend (c : cone F) {X : C} (f : X ⟶ c.X) : cone F :=
```

Comment:
The cone argument should come last in this and most of the below functions. That way it can be usefully partially applied, probably it's a functor too in the appropriate category.

Reply:
Ok, I've changed the order here on `(co)cones.precompose/whisker`. (No subsequent changes were necessary, because Lean is smart!)

I have not made these functors.

Reply:
(I agree that at some point we'll probably want that these are functors, but apparently not yet, so let's get this merged first!)